### ISSUE TITLE
Layz dask working extra just for fun?

### ISSUE THREAD

**Issue by ninahakansson** [pytroll/satpy#2434]
**Describe the bug**
My problem is with processing time. I wanted to read some files and create some images and found processing time suspiciously long. The example  below took 27 seconds. In operation we have a setup with more images and areas and the creation of images could take up to 15 minutes which starts to be a problem.  I did some poor man profiling and found one problem was the _invproj function called in get_lonlats() in geometry.py in pyresample. It takes 3 seconds was evaluated at leas ones per image created. The old way to cache stuff in geometry.py is to be deprecated, but anyway tested to turn it on manually by adding:             

```
              res = map_blocks(_invproj, target_x, target_y,
                             chunks=(2,) + target_x.chunks,
                             meta=np.array((), dtype=target_x.dtype),
                             dtype=target_x.dtype,
                             new_axis=[0], proj_dict=self.crs_wkt)
+            res = res.compute()
              lons, lats = res[0], res[1]
+            self.lons = lons
+            self.lats = lats
```

With this the computation time was reduced to  21s. But as I understand it dask knows better than to recalculate the same array. And the caching with self.lons and self.lats should not be needed?!? To reduce processing time I also want to reuse the same mask and the same resampler for all products produced on the same swath. Here is a suggestion how it could be done: https://github.com/pytroll/satpy/pull/2433. , Now by combining the fix in geometry.py with the PR and adding the mask_array_name="cloudmask" to the resample argument processing time is down to more reasonable 18s. However I expected to still get 21s if I do not use the mask_array_name in my call to resample. But processing time without the new  mask_array_name keyword went up to 42!?! seconds. And with the new PR without the mask_array_name and without the fix in geometry.py the processing is up to over a minute! (The _invproj function is then recalculated 18 times!). 

I run my test with export DASK_NUM_WORKERS=2. I think it is good to use a low number. Otherwise dask can just do a lot of the extra unnecessary work in parallel without effect on processing time. I suspect the _invproj is not the only function to be evaluated too many times, but it is one of the longest so it was easier to find.

The scene.savedatasets() give the same type of problem if used with png or jpg formats.

**To Reproduce**

```python
import time
import glob
from satpy import Scene
from satpy.utils import debug_on
debug_on()
import dask

with dask.config.set(scheduler='threading'):
    tic = time.time()
    filenames = glob.glob("/home/a001865/temp/NWCSAF/*.nc*")
    scn1 = Scene(filenames=filenames, reader = "nwcsaf-pps_nc")
    scn1.load([ 'cloudmask', 'cloudtype', 'cloud_top_phase', ])
    scn2 = scn1.resample("euron1", generate=False)
    a = scn2.save_dataset('cloudmask', 'test_1.png', fill_value=0)
    b = scn2.save_dataset('cloudtype', 'test_2.png', fill_value=0)
    c = scn2.save_dataset('cloud_top_phase', 'test_3.png', fill_value=0)
    a = scn2.save_dataset('cloudmask', 'test_5.jpg', fill_value=0)
    b = scn2.save_dataset('cloudtype', 'test_4.jpg', fill_value=0)
    print(time.time() - tic)
```

**Expected behavior**
All my images produced in around 10 seconds.

**Actual results**
My images produced in 27 seconds.

**Screenshots**

**Environment Info:**
 - OS: [e.g. OSX, Windows, Linux]
 - Satpy Version: [man and 0.41.1]
 - PyResample Version: 1.26.1
 - Readers and writers dependencies (when relevant): [run `from satpy.utils import check_satpy; check_satpy()`]

**Additional context**

**Comment by djhoese** [comment]
So I think these types of optimizations and how they are implemented depend on what you (the user) expect and what you/us (the developers) expect as far as how these interfaces are used. So in your test code you call `save_dataset` for each product you want to save. I expect that this will recompute the longitudes every time. This is because dask isn't "caching" the computations for the entire process, that would waste a lot of memory holding on to those intermediate results. Instead, when you compute something (`.compute()` or `.persist()`) dask will look at everything being computed and figure out what computations can be shared. This is discussed in this part of the Satpy documentation when we want to share dask computations between multiple save calls:

https://satpy.readthedocs.io/en/stable/writers.html#saving-multiple-scenes-in-one-go

Dask does this check by looking at the "task name" of each dask array or Delayed object or whatever other dask object you're using. So for a DataArray with dask underneath you could do `data_arr.data.name` and that should give you the dask task name. If Satpy or Pyresample does some kind of dask operation multiple times (ex. calls `swath_def.get_lonlats()`) and the individual dask arrays have *different* task names when they are the *same* computation on the same data, then this is bad. This means that when dask tries to consolidate/share computations it won't realize these results are the same thing. Now this could be caused by the input arrays (ex. being read/created by the reader) having different names when they are the same thing, or this could be `map_blocks` or other dask-related calls not being configured correctly.

I think in Pyresample 2.0 (the redesign of pyresample I'm working on), rather than having internal caching in the `SwathDefinition`, I'd rather have convenience methods like `.persist()` or `.compute()` that create a new SwathDefinition from the dask-based one. This is more difficult on the Satpy-side of things since it means swaths have to be explicitly converted, but it keeps the pyresample stuff very simple.

**Comment by ninahakansson** [comment]
I rerun my experiment using the compute_writer_results. See the example below:
Without any of my changes I got 6 images in 32 seconds. With the added caching in pyresample geometry I got it reduced to 24 seconds. And with my PR for reusing the remapper it is down to 12s. 

```
import time
import glob
from satpy import Scene
from satpy.utils import debug_on
#debug_on()
import dask
from satpy.writers import compute_writer_results

with dask.config.set(scheduler='threading'):
    tic = time.time()
    filenames = glob.glob("/home/a001865/temp/NWCSAF/*.nc*")
    scn1 = Scene(filenames=filenames, reader = "nwcsaf-pps_nc")
    scn1.load([ 'cloudtype', 'cloudmask', "cloud_top_phase"])
    scn2 = scn1.resample("euron1", compute=False)
    scn2.unload()
    res1 = scn2.save_datasets(filename="/home/a001865/temp/{name}.png",
                              datasets = ["cloudmask", "cloudtype", "cloud_top_phase"],
                              writer='simple_image',
                              compute=False)

    res2 = scn2.save_datasets(filename="/home/a001865/temp/hej{name}.jpg",
                              datasets = ["cloudmask", "cloudtype", "cloud_top_phase"],
                              fill_value = 0,
                              writer='simple_image',
                              compute=False)

    results = [res1, res2]
    try:
        compute_writer_results(results)
    except ValueError:
        # ?? ValueError: Different number of sources [0] and targets [6]"
        pass
    print(time.time() - tic)
 ```

**Comment by djhoese** [comment]
Just for comparison, with your above code, what do you get if you don't do `compute_writer_results` and leave `save_datasets(compute=True)`?

Do I understand correctly that the reuse of resamplers in your PR is basically saying "these swath masks are the same so we can reuse the same KDTree", right? I'm not sure if I understand how this all works or how widely it can be used. I'm intrigued by the performance differences you're seeing, but I still worry that any performance issues you are seeing are tied to dask bugs in the reader you're using or are made worse by saving to png/jpg (not dask chunk-friendly). Lastly, related to this, are performance issues originally discovered by @gerritholl when dask delayed objects are mixed with dask arrays...but that's a larger issue possibly outside the scope of this issue/PR.

**Comment by ninahakansson** [comment]
I got the same time (34s) with savedatasets and compute=True.

**Comment by djhoese** [comment]
I thought I commented this already, but now I don't see it. Do you have some files you could send me so I could look into any possible issues with this reader?

> I got the same time (34s) with savedatasets and compute=True

This tells me something is not being reused when it should be. We should be seeing at least some improvement between individual `save_datasets` calls and a combined `save_datasets`. This makes me think that the dask task names are not correct in some way.

**Comment by djhoese** [comment]
So the good news is that nothing extremely obvious is wrong with the swath definitions:

```
In [9]: scn['cloudtype'].attrs['area'].lons.data.name
Out[9]: 'add-9269fb81e9d6fcec21bca4fd92c1557e'

In [10]: scn['cloudmask'].attrs['area'].lons.data.name
Out[10]: 'add-9269fb81e9d6fcec21bca4fd92c1557e'

In [11]: scn['cloud_top_phase'].attrs['area'].lons.data.name
Out[11]: 'add-9269fb81e9d6fcec21bca4fd92c1557e'

In [12]: scn['cloudtype'].attrs['area'].lats.data.name
Out[12]: 'add-7dcc3274c9c772f6193bce48b43c37fe'

In [13]: scn['cloudmask'].attrs['area'].lats.data.name
Out[13]: 'add-7dcc3274c9c772f6193bce48b43c37fe'

In [14]: scn['cloud_top_phase'].attrs['area'].lats.data.name
Out[14]: 'add-7dcc3274c9c772f6193bce48b43c37fe'
```

I do notice that with the default chunk size of 4096 that this may be performing worse than it needs to. Have you tried setting `PYTROLL_CHUNK_SIZE` to 1024 or something similar to see how it effects timing?

I'll keep looking at some of the stuff you're seeing.

**Comment by djhoese** [comment]
I took your original example but left `generate` at the default and then added a `print` statement to `_invproj` in geometry.py. I only see one computation:

```python
In [1]: from satpy import Scene; from glob import glob
scn
In [2]: scn = Scene(reader='nwcsaf-pps_nc', filenames=glob("/data/satellite/NWCSAF/*.nc"))

In [3]: scn.load(['cloudtype', 'cloudmask', 'cloud_top_phase'])
Can't load ancillary dataset cma_conditions
Can't load ancillary dataset cma_quality
Can't load ancillary dataset cma_status_flag

In [4]: new_scn = scn.resample("euron1")
/home/davidh/repos/git/pyresample/pyresample/geometry.py:695: FutureWarning: This function is deprecated. See: https://pyproj4.github.io/pyproj/stable/gotchas.html#upgrading-to-pyproj-2-from-pyproj-1
  xyz = np.stack(transform(src, dst, lons, lats, alt), axis=1)

In [5]: new_scn.compute()
/home/davidh/miniconda3/envs/satpy_py310/lib/python3.10/site-packages/pyproj/crs/crs.py:1296: UserWarning: You will likely lose important projection information when converting to a PROJ string from another format. See: https://proj.org/faq.html#what-is-the-best-format-for-describing-coordinate-reference-systems
  proj = self._crs.to_proj4(version=version)
Running _invproj (3072, 3072)
Out[5]: <satpy.scene.Scene at 0x7f9e341034f0>
```

You were saying it gets called 18ish times? I just tried `generate=False` and see the same results. What is different between our environments?

**Comment by ninahakansson** [comment]
I updated my env and got better performance. The The _inproj was then called only ones during resampling. 
Old env had:
```
python                    3.9.15          hba424b6_0_cpython    conda-forge
dask                      2023.1.0           pyhd8ed1ab_0    conda-forge
xarray                    2023.1.0           pyhd8ed1ab_0    conda-forge
pyresample                1.26.1                   pypi_0    pypi
```

New env:
```
python                    3.11.2          h2755cc3_0_cpython    conda-forge
dask                      2023.3.2           pyhd8ed1ab_0    conda-forge
xarray                    2023.3.0           pyhd8ed1ab_0    conda-forge
pyresample                1.26.1          py311h2872171_0    conda-forge
```

**Comment by djhoese** [comment]
Unless it was a bug in dask, I'm surprised that these version updates would change the number of times _invproj got called. I'd expect Python 3.11 to give us a little performance improvement, but not to change dask/xarray behavior.

**Comment by ninahakansson** [comment]
I tried several combinations of packages to create a new but bad environment. But every combination was just as good as the new one. Absolutely I could have broken something in my environment but we also built several environments in production with very slow generation of the images (python 3.9.7). I tested a new one with python 3.11 today and generation of images was much faster. I will leave the issue open a few days but it seem nothing is actually broken any longer :).

**Comment by djhoese** [comment]
Sounds good. Yeah, if we can get reproducible examples of dask calculating something multiple times when we don't expect it then absolutely we need to investigate it.
